packages/devextreme/js/__internal/core/state_manager/state_manager.ts

```diff
@@ -0,0 +1,270 @@
+import type {
+  IController,
+  IControllerRegistry,
+  IDevToolsConnector,
+  ILogger,
+  IState,
+  IStateChange,
+  IStateContainer,
+  IStateHistory,
+  IStateManager,
+  IStateManagerActionType,
+  IStateManagerConfig,
+  IStateTracker,
+} from './interfaces';
+import { isValidStatePath, splitStatePath } from './utils';
+
+export class StateManager implements IStateManager {
+  private readonly controllerRegistry: IControllerRegistry;
+
+  private readonly stateTracker: IStateTracker;
+
+  private readonly stateHistory: IStateHistory;
+
+  private readonly devToolsConnector: IDevToolsConnector;
+
+  private readonly logger: ILogger;
+
+  private isRestoring = false;
+
+  private isWarnedAboutUnstableJumpToStateFunctionality = false;
+
+  constructor(
+    config: IStateManagerConfig,
+  ) {
+    const {
+      controllerRegistry, stateTracker, stateHistory, devToolsConnector, logger,
+    } = config;
+
+    this.controllerRegistry = controllerRegistry;
+    this.stateTracker = stateTracker;
+    this.stateHistory = stateHistory;
+    this.devToolsConnector = devToolsConnector;
+    this.logger = logger;
+
+    this.setupEventHandlers();
+    this.devToolsConnector.connect();
+
+    this.logger.info('StateManager initialized');
+  }
+
+  registerController(...args: Parameters<IControllerRegistry['registerController']>): ReturnType<IControllerRegistry['registerController']> {
+    const [controller, id] = args;
+    return this.controllerRegistry.registerController(controller, id);
+  }
+
+  restoreState(state: IState): void {
+    try {
+      this.isRestoring = true;
+      const controllers = this.controllerRegistry.getAllControllers();
+
+      Object.keys(state).forEach((controllerId) => {
+        const controller = controllers[controllerId];
```

Comment:
Why not using the absraction method?
Using entries outside the smart collection breaks incapsulation.

```suggestion
        const controller = this.controllerRegistry.getController(controllerId);
```

Reply:
If there is a need of making a snapshot of the `controllerRegistry` at specific moment, then we should review the case, as it's very unclear and hides some concurrency problems.

In this case we might say that `ControllerRegistry`can't serve its purpose well

Reply:
As I mentioned before, it is an unstable feature, so I added a warning message.

[`The state travel feature is unstable and may cause unexpected behavior.`](https://github.com/DevExpress/DevExtreme/blob/8ec43dd1251ed00e46849a34708d7526f09f1526/packages/devextreme/js/__internal/core/state_manager/state_manager.ts#L226-L230)

My goal is to collect feedback about the necessity of this feature.

If consumers try it and see that it would be great to have, I will improve it; if not, I will remove it.

Reply:
The `getAllControllers` method breaks the abstraction — that’s why the question arises.
It's about `ControllerRegistry` responsibilities

Reply:
Removed the controller registry